File: 2022/BOJ/14. BackTracking/9663.cpp
#include <iostream>
using namespace std;

#define MAX 15

int board[MAX];
int answer;
int n;

bool is_pormising(int x)
{
    for (int i = 0; i < x; i++)
        if (board[i] == board[x] || abs(x - i) == abs(board[x] - board[i]))
            return false;

    return true;
}

void n_queen(int x)
{
    if (x == n)
    {
        answer += 1;
        return;
    }

    for (int i = 0; i < n; i++)
    {
        board[x] = i;
        if (is_pormising(x))
            n_queen(x + 1);
    }
}

int main(void)
{
    cin >> n;

    n_queen(0);
    cout << answer << "\n";

    return 0;
}

File: 2022/BOJ/14. BackTracking/9663.py
import sys

n = int(sys.stdin.readline())

answer = 0
board = [0] * n


def n_queen(x: int):
    global answer

    def is_promising(x: int):
        for i in range(x):
            if board[x] == board[i] or abs(x - i) == abs(board[i] - board[x]):
                return False

        return True

    if x == n:
        answer += 1
        return

    for i in range(n):
        board[x] = i
        if is_promising(x):
            n_queen(x + 1)


n_queen(0)
print(answer)
